### PR TITLE
Implementation of the folder navigation by URL 

### PR DESCRIPTION
**Pull request by Oleh-Stefanovych**
Now the current folder path is displayed in the URL and handles its change.

The `ESLint` result:
image

### CODE REVIEW

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, PTAL.

File: app/presenter/files-presenter/index.js [hunk 1]
```diff
@@ -15,21 +15,42 @@ export class FilesPresenter {
...
+   * Loads a list of files in the specified folder. If the target folder id is not specified, it
+   * gets the root folder id using the request.
+   * @param {string} folderId - a tarhet folder id.
+   * @private
+   */
+  _loadFolder(folderId = undefined) {
+    if (folderId) {
+      this.store.dispatch('getFolderById', folderId)
+          .then((folder) => {
+            return this._filesPageComponent.loadFolder(folder);
```
[the remark this answers is not in the document]
Comment (Oleh-Stefanovych) [reply app/presenter/files-presenter/index.js]: - FilePresenter is responsible for loading the folder into which files should be displayed.
- FilePage is responsible for downloading only files.

Accordingly, FilePresenter tells FilePage which folder to load.

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, fixed, PTAL.

Fixes:
- Replaced state setting with mutator.
- Removed unnecessary default parameter value.

`Is this PR finalized?` - Yes. Event handling for navigation between components was implemented [here](https://github.com/Oleh-Stefanovych/FileStorage-UI/pull/18).

File: app/store/actions.js [hunk 2]
```diff
@@ -138,4 +141,35 @@ export default {
     };
   },
 
+  getFolderById(store) {
```
Comment (dkalinin) [inline app/store/actions.js]: Let's rename.

File: app/store/actions.js [hunk 3]
```diff
@@ -138,4 +141,35 @@ export default {
...
+            return JSON.parse(response)['folder'];
+          }).catch((error) => {
+            throw new Error(JSON.parse(error.message)['errorMessage']);
+          }).finally(() => {
+            store.commit('clearFileListStatus');
+          });
+    };
+  },
+
+  defineFiles(store) {
```
Comment (dkalinin) [inline app/store/actions.js]: Should be done in init state.

File: app/store/actions.js [hunk 4]
```diff
@@ -138,4 +141,35 @@ export default {
...
+    };
+  },
+
+  showServerMessage(store) {
+    return (serverMessage) => {
+      store.commit('setServerMessage', serverMessage);
+    };
+  },
+
+  setFileListControlsEnabling(store) {
```
Comment (dkalinin) [inline app/store/actions.js]: Please remove.

File: app/store/actions.js [hunk 5]
```diff
@@ -138,4 +141,35 @@ export default {
...
+    };
+  },
+
+  defineFiles(store) {
+    return () => {
+      store.commit('setFiles', []);
+    };
+  },
+
+  showServerMessage(store) {
```
Comment (dkalinin) [inline app/store/actions.js]: This action is likely replaced by commit. Server messages are handled in actions.

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, fixed, PTAL.

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, fixed, PTAL.

Fixes:
- Redirect if URL is "/"
- Set current folder if URL is '/:id'
- Fetch files using displayed folder from state.
- Create separate fields for the different server errors.
- Refactor `Router` class.

File: app/component/high-order-component/files-page/index.js [hunk 6]
```diff
@@ -34,24 +34,71 @@ export class FilesPage extends HighOrderComponent {
...
+      this._uploadFilesButton.enabled = false;
+      alert(serverMessage);
+    });
+
+    this.store.eventBus.subscribe('stateChanged.serverError.uploadFiles', (serverMessage) => {
+      alert(serverMessage);
+    });
+
+    this.store.eventBus.subscribe('stateChanged.serverError.moveFile', (serverMessage) => {
+      alert(serverMessage);
```
Comment (dkalinin) [inline app/component/high-order-component/files-page/index.js]: How about to replace alerts with nice error messages?
https://getbootstrap.com/docs/3.4/components/#alerts

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, PTAL:

- Implementation of the folder navigation by click was implemented in #18.
- Implementation of the server message display in the style of [bootstrap ](https://getbootstrap.com/docs/3.4/components/#alerts) was implemented in #27.

File: app/router/index.js [hunk 7]
```diff
@@ -25,6 +34,24 @@ export class Router {
...
 
+  /**
+   * Returns an array of {@link RoutingConfigEntry}, in each of which the path has only the static
+   * part, excluding the dynamic part.
+   * @return {Array<RoutingConfigEntry>} - an array of configuration entries, in each of which
+   * the path has only the static part, excluding the dynamic part.
+   */
+  get staticRoutingConfigEntries() {
+    const staticRoutingConfig = [];
+    this._routingConfig.forEach((routingEntry) => {
```
Comment (dkalinin) [inline app/router/index.js]: Please use JS Array .map()

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, PTAL.

Fixes:
- Removed redundant catch blocks.
- Removed redundant error throwing
- Replaced `forEach` block witch `Array.map()`.

**Note**:
- Implementation of the folder navigation by click was implemented in #18. 
- Implementation of the server message display in the style of bootstrap was implemented in #27.
- Implementation of the server error handling was imlemented in #29.

File: app/router/index.js [hunk 8]
```diff
@@ -41,15 +41,12 @@ export class Router {
...
-    this._routingConfig.forEach((routingEntry) => {
-      staticRoutingConfig.push({
-        path: routingEntry['path'].split(':')[0],
-        presenter: routingEntry['presenter'],
-      });
+    return this._routingConfig.map((routingEntry) => {
+      const newRoutingEntry = {};
+      newRoutingEntry.path = routingEntry.path.split(':')[0];
+      newRoutingEntry.presenter = routingEntry.presenter;
+      return newRoutingEntry;
```
Comment (dkalinin) [inline app/router/index.js]: ES6 trick:
```js
return this._routingConfig.map(({path, presenter}) => ({
  path: path.split(':')[0],
  presenter
})
```
Not tested!

**Pull request merged**